cowbird/typedefs.py

```diff
@@ -3,63 +3,143 @@
-    _JsonItem = Union[AnyValueType, _JsonObjectItem, _JsonListItem, _JSON]
-    JSON = Union[Dict[str, _JsonItem], List[_JsonItem], AnyValueType]
+ParamsType = Dict[str, Any]
+CookiesType = Union[Dict[str, str], List[Tuple[str, str]]]
+HeadersType = Union[Dict[str, str], List[Tuple[str, str]]]
+AnyHeadersType = Union[HeadersType, ResponseHeaders, EnvironHeaders, CaseInsensitiveDict]
+AnyCookiesType = Union[CookiesType, RequestsCookieJar]
+AnyResponseType = Union[WebobResponse, PyramidResponse, RequestsResponse, HTTPException, TestResponseType]
 
-    # registered configurations
-    ConfigItem = Dict[str, JSON]
-    ConfigList = List[ConfigItem]
-    ConfigDict = Dict[str, Union[str, ConfigItem, ConfigList, JSON]]
-    ConfigResTokenInfo = TypedDict("ConfigResTokenInfo", {"has_multi_token": bool, "named_tokens": set})
-    ConfigSegment = TypedDict("ConfigSegment", {"name": str, "type": str})
+# pylint: disable=C0103,invalid-name
+ValueType = Union[str, Number, bool]
+AnyValueType = Optional[ValueType]  # avoid naming ambiguity with PyWPS AnyValue
+AnyKey = Union[str, int]
+# add more levels of explicit definitions than necessary to simulate JSON recursive structure better than 'Any'
+# amount of repeated equivalent definition makes typing analysis 'work well enough' for most use cases
+_JSON: TypeAlias = "JSON"
+_JsonObjectItemAlias: TypeAlias = "_JsonObjectItem"
+_JsonListItemAlias: TypeAlias = "_JsonListItem"
+_JsonObjectItem = MutableMapping[str, Union[_JSON, _JsonObjectItemAlias, _JsonListItemAlias, AnyValueType]]
+_JsonListItem = MutableSequence[Union[_JsonObjectItem, _JsonListItemAlias, AnyValueType]]
+_JsonItem = Union[_JsonObjectItem, _JsonListItem, AnyValueType]
+# NOTE:
+#   Although 'JSON' should allow referring directly to anything between 'Dict[str, JSON]', 'List[JSON]'
+#   or 'AnyValueType', this can a lot of false positives typing detections. Sometimes, it is better to provide
+#   the explicit type expected (e.g.: 'List[JSON]') when necessary to disambiguate some situations.
+JSON = Union[MutableMapping[str, _JsonItem], MutableSequence[_JsonItem], AnyValueType]
 
-    ResourceSegment = TypedDict("ResourceSegment", {"resource_name": str, "resource_type": str})
-    PermissionDataEntry = TypedDict("PermissionDataEntry",
-                                    {"res_path": List[ResourceSegment], "permissions": Dict[str, List[str]]})
-    PermissionData = Dict[str, PermissionDataEntry]
+HTTPMethod = Literal[
+    "HEAD",
+    "GET",
+    "PUT",
+    "POST",
+    "PATCH",
+    "DELETE",
+    # others available, but not common for API definition
+]
 
-    from cowbird.database.stores import StoreInterface
-    StoreSelector = Union[Type[StoreInterface], StoreInterface, str]
+HandlerConfig = TypedDict(
+    "HandlerConfig",
+    {
+        "active": Optional[bool],
+        "priority": Optional[int],
+        "url": Optional[str],
+        "workspace_dir": Optional[str],
```

Comment:
Yes, we can add them as well. Are those the only other possible names?

Reply:
Yes, for now. More will be added in my next PR. Note that they are not used in all handlers, but I think it's fine since they are added as `Optional`.

Reply:
I'm about to push a new commit. I'm using `NotRequired` that better represents which fields are mandatory or not.